hardhat/README.md

```diff
@@ -0,0 +1,236 @@
+
+1. Open a second command line terminal
+2. Run `hardhat`, the development environment we use in this tutorial
+   ```sh
+   mkdir dapp
+   cd dapp
+   npx hardhat
+   ```
+3. Select **Create a sample project** and accept all the defaults.
+4. Verify the sample application.
+   ```sh
+   npx hardhat test
+   ```
+   
+#### Interact with the Sample App Manually (optional)   
+   
+If you want to be more hands on, you can interact with the contract manually.
+
+1. Start the console
+   ```sh
+   npx hardhat console
+   ```
+2. Deploy the greeter contract.
+   ```javascript
+   const Greeter = await ethers.getContractFactory("Greeter")
+   const greeter = await Greeter.deploy("Hello, world!")
+   await greeter.deployed()
+   ```
+3. Get the current greeting.
+   ```javascript
+   await greeter.greet()
+   ```
+4. Modify the greeting.
+   ```javascript
+   const tx = await greeter.setGreeting("Hola, mundo")
+   await tx.wait()
+   ```
+5. Verify the greeting got modified.
+   ```javascript
+   await greeter.greet()
+   ```
+   
+6. Leave the console.
+   ```javascript
+   .exit
+   ```
+
+### Migrate the Sample App to Optimistic Ethereum
+
+Now that we have a running Optimistic Ethereum server, and a dapp to run on it, we can run on Optimistic Ethereum.
+
+1. Install the Optimistic Ethereum package in the application.
+   ```sh
+   yarn add @eth-optimism/hardhat-ovm
+   ```
+2. Edit `hardhat.config.js` to use the Optimistic Ethereum package.
+   ```js
+   require("@nomiclabs/hardhat-waffle");
+   require('@eth-optimism/hardhat-ovm')
+
+   ...
+   ```
+3. In the same file, add `optimistic` to the list of networks:
+   ```js
+   // hardhat.config.js
+
+   ...
+   
+   module.exports = {
+     solidity: "0.8.4",   // or whatever you get from HardHat
+     networks: {
+       // Add this network to your config!
+       optimistic: {
+          url: 'http://127.0.0.1:8545',
+          accounts: {
+             mnemonic: 'test test test test test test test test test test test junk'
+          },
+          // This sets the gas price to 0 for all transactions on L2. We do this
+          // because account balances are not automatically initiated with an ETH
+          // balance (yet, sorry!).
+          gasPrice: 0,
+          ovm: true // This sets the network as using the ovm and ensure contract will be compiled against that.
+       }
+     }
+   }
+   ```
+
+   At this point you need to wait until the `docker-compose build` ends, if it hasn't yet, and then run
+   `cd ~/optimism/ops ; docker-compose up`.
+
+4. Test the contract on Optimistic Ethereum. Hardhat will recognize it has not been compiled and compile it for you.
+
+   ```sh
+   npx hardhat --network optimistic test
+   ```
+
+5. If you want to interact with the app manually, use the console. You can use the same JavaScript commands
+   to control it you used above.
+   ```sh
+   npx hardhat --network optimistic console
+   ```
+   
+   
+#### Change the Solidity Version if Needed
```

Comment:
Hmmm. Do current versions of hardhat always give 0.8.0+? If so, maybe worth explaining this as soon as the user runs `npx hardhat` instead of here since they might run into this bug and be confused when running the above commands.

Reply:
The same program is for 0.8.4 (IIRC). But the above commands will all work, this is where it becomes a problem. Also, doing it here lets me illustrate that when you change anything you need to test it first in Ethereum and only then in Optimistic Ethereum.